test/Microsoft.AspNetCore.Sockets.Common.Tests/Internal/Formatters/ServerSentEventsParserTests.cs

```diff
@@ -0,0 +1,234 @@
+            var consumed = new ReadCursor();
+            var examined = new ReadCursor();
+
+            var parsePhase = parser.ParseMessage(readableBuffer, out consumed, out examined, out Message message);
+            Assert.Equal(ServerSentEventsMessageParser.ParseResult.Completed, parsePhase);
+
+            var result = Encoding.UTF8.GetString(message.Payload);
+            Assert.Equal(expectedMessage, result);
+        }
+
+        [Theory]
+        [InlineData("data: X\r\n", "Unknown message type: 'X'")]
+        [InlineData("data: T\n", "A '\\n' character can only be used as a line ending")]
+        [InlineData("data: X\r\n\r\n", "Unknown message type: 'X'")]
+        [InlineData("data: Not the message type\r\n\r\n", "Unknown message type: 'N'")]
+        [InlineData("data: T\r\ndata: Hello, World\r\r\n\n", "There was an error in the frame format")]
+        [InlineData("data: Not the message type\r\r\n", "Unknown message type: 'N'")]
+        [InlineData("data: T\r\ndata: Hello, World\n\n", "A '\\n' character can only be used as a line ending")]
+        [InlineData("data: T\r\nfoo: Hello, World\r\n\r\n", "Expected the message prefix 'data: '")]
+        [InlineData("foo: T\r\ndata: Hello, World\r\n\r\n", "Expected the message prefix 'data: '")]
+        [InlineData("food: T\r\ndata: Hello, World\r\n\r\n", "Expected the message prefix 'data: '")]
+        [InlineData("data: T\r\ndata: Hello, World\r\n\n", "There was an error in the frame format")]
+        [InlineData("data: T\r\ndata: Hello\n, World\r\n\r\n", "A '\\n' character can only be used as a line ending")]
+        [InlineData("data: data: \r\n", "Unknown message type: 'd'")]
+        public void ParseSSEMessageFailureCases(string encodedMessage, string expectedExceptionMessage)
+        {
+            var buffer = Encoding.UTF8.GetBytes(encodedMessage);
+            var readableBuffer = ReadableBuffer.Create(buffer);
+            var parser = new ServerSentEventsMessageParser();
+            var consumed = new ReadCursor();
+            var examined = new ReadCursor();
+
+            var ex = Assert.Throws<FormatException>(() => { parser.ParseMessage(readableBuffer, out consumed, out examined, out Message message); });
+            Assert.Equal(expectedExceptionMessage, ex.Message);
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData("data:")]
+        [InlineData("data: \r")]
+        [InlineData("data: T\r\nda")]
+        [InlineData("data: T\r\ndata:")]
+        [InlineData("data: T\r\ndata: Hello, World")]
+        [InlineData("data: T\r\ndata: Hello, World\r")]
+        [InlineData("data: T\r\ndata: Hello, World\r\n")]
+        [InlineData("data: T\r\ndata: Hello, World\r\n\r")]
+        [InlineData("data: T\r\ndata: Hello, World\r\n\r\\")]
+        [InlineData("data: T\r\ndata: Major\r\ndata:  Key\rndata:  Alert\r\n\r\\")]
+        [InlineData("data: T\r\ndata: Major\r\ndata:  Key\r\ndata:  Alert\r\n\r\\")]
```

Comment:
Is `\\` this correct/intentional? `\n` is not two characters so if the intention was that the first message ended with `\` and the last message was just `n` that it is incorrect. 
This actually might be a bug in the parser if this passes - after `\r\n` the only two expected things are `\r\n`  (end of message) or `data: ` (more data). Since you already parsed `\r` after '\r\n` the only valid value is `\n` but in the last two tests you allow `\`